src/proto/pi/mp.rs

```diff
@@ -0,0 +1,199 @@
+}
+
+/// Information about physical location of the processor.
+#[repr(C)]
+#[derive(Default, Debug)]
+pub struct CPUPhysicalLocation {
+    /// Zero-based physical package number that identifies
+    /// the cartridge of the processor.
+    pub package: u32,
+    /// Zero-based physical core number within package of the processor.
+    pub core: u32,
+    /// Zero-based logical thread number within core of the processor.
+    pub thread: u32,
+}
+
+/// Protocol that provides services needed for multi-processor management.
+#[repr(C)]
+#[unsafe_guid("3fdda605-a76e-4f46-ad29-12f4531b3d08")]
+#[derive(Protocol)]
+pub struct MPServices {
+    get_number_of_processors: extern "win64" fn(
+        this: *const MPServices,
+        number_of_processors: *mut usize,
+        number_of_enabled_processors: *mut usize,
+    ) -> Status,
+    get_processor_info: extern "win64" fn(
+        this: *const MPServices,
+        processor_number: usize,
+        processor_info_buffer: *mut ProcessorInformation,
+    ) -> Status,
+    startup_all_aps: extern "win64" fn(
+        this: *const MPServices,
+        procedure: Procedure,
+        single_thread: bool,
+        wait_event: *mut c_void,
+        timeout_in_micro_seconds: usize,
+        procedure_argument: *mut c_void,
+        failed_cpu_list: *mut *mut usize,
+    ) -> Status,
+    startup_this_ap: extern "win64" fn(
+        this: *const MPServices,
+        procedure: Procedure,
+        processor_number: usize,
+        wait_event: *mut c_void,
+        timeout_in_micro_seconds: usize,
+        procedure_argument: *mut c_void,
+        finished: *mut bool,
+    ) -> Status,
+    switch_bsp: extern "win64" fn(
+        this: *const MPServices,
+        processor_number: usize,
+        enable_old_bsp: bool,
+    ) -> Status,
+    enable_disable_ap: extern "win64" fn(
+        this: *const MPServices,
+        processor_number: usize,
+        enable_ap: bool,
+        health_flag: *const u32,
+    ) -> Status,
+    who_am_i: extern "win64" fn(this: *const MPServices, processor_number: *mut usize) -> Status,
+}
+
+impl MPServices {
+    /// Retrieves the number of logical processors and the number of enabled logical processors in the system.
+    pub fn get_number_of_processors(&self) -> Result<ProcessorCount> {
+        let mut total: usize = 0;
+        let mut enabled: usize = 0;
+        (self.get_number_of_processors)(self, &mut total, &mut enabled)
+            .into_with_val(|| ProcessorCount { total, enabled })
+    }
+
+    /// Gets detailed information on the requested processor at the instant this call is made.
+    pub fn get_processor_info(&self, processor_number: usize) -> Result<ProcessorInformation> {
+        let mut pi: ProcessorInformation = Default::default();
+        (self.get_processor_info)(self, processor_number, &mut pi).into_with_val(|| pi)
+    }
+
+    /// Executes provided function on all APs in blocking mode.
+    pub fn startup_all_aps(
```

Comment:
Do you mean taking `T` and calling `mem::transmute()` on it inside of `startup_all_aps`/`startup_this_ap`?

Also, I thought of making something like this:

https://github.com/rust-osdev/uefi-rs/blob/2bc22bbdeca270a881dc1e08822af922a5b9b659/src/table/boot.rs#L272-L276

To make it possible to use ordinary Rust functions that will get an user-specified parameter as their argument (parametrized via `startup_all_aps`/`startup_this_ap`).

But this would require parametrizing that FFI function too, which isn't possible.

Reply:
Yeah, having the generic function call `mem::transmute` is what I had in mind

Reply:
OK, just need to clarify one thing:

If you look at the typical procedure (`proc_increment_atomic()` or `proc_wait_100ms()` in tests), the user will still need to call `mem::transmute()` and correctly specify the original type to get the argument back.

So now we have one magically converted argument that's not magically gets converted back, and this makes things unbalanced as you can't just audit your code by looking for matching `mem::transmute()` pairs (one on the "sender" side, and the other on the "receiver" side).

So even through this might indeed make calling `startup_all_aps`/`startup_this_ap` more ergonomic, the overall usability/auditability might get hurt. You still think we should proceed this way?

Reply:
You're right, it wouldn't help. When fixing the clippy lints I've discovered that the transmutes when calling the function can be replaced with safe pointer casts.